cluster/images/etcd/migrate/data_dir.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang/glog"
+)
+
+// DataDirectory provides utilities for initializing and backing up an
+// etcd "data-dir" as well as managing a version.txt file to track the
+// etcd server version and storage verion of the etcd data in the
+// directory.
+type DataDirectory struct {
+	path        string
+	dir         *os.File
+	versionFile *VersionFile
+}
+
+// OpenOrCreateDataDirectory opens a data directory, creating the directory
+// if it doesn't not already exist.
+func OpenOrCreateDataDirectory(path string) (*DataDirectory, error) {
+	exists, err := exists(path)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		glog.Infof("data directory '%s' does not exist, creating it", path)
+		err := os.MkdirAll(path, 0755)
```

Comment:
I realize that the script your replacing was unspecific on the permissions for the directory being created. I believe we were probably creating it with 755 but it seems that the data directory containing things like secrets should not be world readable.

Reply:
Thanks for the nudge. I'll review what exactly happens in practice (in GCE/GKE) and see if we can restrict this to 0750

Reply:
Related question: does our umask impact this?  (I think yes.)  By default, GNU `mkdir` with no arguments uses `0777` (modified by umask).  So the "no change" change would be to use `0777` and probably add a comment asking if we can/should tighten that up.

Unexpectedly *restricting* the permissions on update could cause an outage.